OpenRA.Mods.Common/EditorBrushes/EditorTileBrush.cs

```diff
@@ -106,8 +123,66 @@ public bool HandleMouseInput(MouseInput mi)
 					}
 				}
 			}
+		}
 
-			return true;
+		void FloodFillWithBrush(CPos cell, bool isMoving)
+		{
+			var map = world.Map;
+			var mapTiles = map.MapTiles.Value;
+			var replace = mapTiles[cell];
+
+			if (replace.Type == Template)
+				return;
+
+			var queue = new Queue<CPos>();
+			var touched = new bool[map.MapSize.X, map.MapSize.Y];
+
+			var rules = map.Rules;
+			var tileset = rules.TileSets[map.Tileset];
+			var template = tileset.Templates[Template];
+
+			Action<CPos> maybeEnqueue = (newCell) =>
+			{
+				if (map.Contains(cell) && !touched[newCell.X, newCell.Y])
+				{
+					queue.Enqueue(newCell);
+					touched[newCell.X, newCell.Y] = true;
+				}
+			};
+
+			queue.Enqueue(cell);
+			while (queue.Count > 0)
+			{
+				var queuedCell = queue.Dequeue();
+				if (mapTiles[queuedCell].Type != replace.Type)
+					continue;
+
+				var previousCell = FindEdge(map, queuedCell, new CVec(-1 * template.Size.X, 0), replace);
+				var nextCell = FindEdge(map, queuedCell, new CVec(1 * template.Size.X, 0), replace);
+
+				for (var x = previousCell.X; x <= nextCell.X; x += template.Size.X)
+				{
+					PaintCell(new CPos(x, queuedCell.Y), isMoving);
+					var upperCell = new CPos(x, queuedCell.Y - (1 * template.Size.Y));
```

Comment:
This looks a bit iffy to me.  Why aren't you just looping over all the cells in the template directly?